src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -397,6 +370,125 @@ public void FinishQuerying()
 			}
 		}
 
+		private async Task<PayloadData> SendClearPasswordAsync(ConnectionSettings cs, IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			// add NUL terminator to password
+			var passwordBytes = Encoding.UTF8.GetBytes(cs.Password);
+			Array.Resize(ref passwordBytes, passwordBytes.Length + 1);
+
+			// send plaintext password
+			var payload = new PayloadData(new ArraySegment<byte>(passwordBytes));
+			await SendReplyAsync(payload, ioBehavior, cancellationToken).ConfigureAwait(false);
+			return await ReceiveReplyAsync(ioBehavior, cancellationToken).ConfigureAwait(false);
+		}
+
+#if !NET45
+		private async Task<PayloadData> SendEncryptedPasswordAsync(
+			string rsaPublicKey,
+			ConnectionSettings cs,
+			IOBehavior ioBehavior,
+			AuthenticationMethodSwitchRequestPayload switchRequest,
+			CancellationToken cancellationToken)
+		{
+			// load the RSA public key
+			RSA rsa;
+			try
+			{
+				rsa = Utility.DecodeX509PublicKey(rsaPublicKey);
+			}
+			catch (Exception ex)
+			{
+				throw new MySqlException("Couldn't load server's RSA public key; try using a secure connection instead.", ex);
+			}
+
+			// add NUL terminator to password
+			var passwordBytes = Encoding.UTF8.GetBytes(cs.Password);
+			Array.Resize(ref passwordBytes, passwordBytes.Length + 1);
+
+			using (rsa)
+			{
+				// XOR the password bytes with the challenge
+				AuthPluginData = Utility.TrimZeroByte(switchRequest.Data);
+				for (var i = 0; i < passwordBytes.Length; i++)
+					passwordBytes[i] ^= AuthPluginData[i % AuthPluginData.Length];
+
+				// encrypt with RSA public key
+				var encryptedPassword = rsa.Encrypt(passwordBytes, RSAEncryptionPadding.Pkcs1);
```

Comment:
I have changed RSA padding from OaepSHA1 to Pkcs1, it may have caused this failed test `SideBySide.ConnectAsync.Sha256WithoutSecureConnection [FAIL]` (https://travis-ci.org/mysql-net/MySqlConnector/jobs/283268492).

Reply:
The [MySQL documentation for `sha256_password`](https://web.archive.org/web/20170404163410/https://dev.mysql.com/doc/internals/en/sha256.html) says:

> RSA encrypts that with the server's public key using `RSA_PKCS1_OAEP_PADDING` as padding

I found that I needed to use `RSAEncryptionPadding.OaepSHA1` to be compatible.

Reply:
It didn't work with `caching_sha2_password` so I have switched it and all tests passed locally.

Reply:
As an experiment, I changed the padding to be different for each auth method and the tests passed: fb884aa5080c248e33f840f455a45c52d96279b1

However, I'm unsettled that it would need to be different for each case, and I'd like to find proof from the documentation (or source code) before committing to that.

Reply:
It is actually different for the two authentication methods.

`sha256_password` uses `RSA_PKCS1_OAEP_PADDING`: https://github.com/mysql/mysql-server/blob/020025d9f277fd3283b8c75024f590857f58fbbd/sql-common/client_authentication.cc#L267

`caching_sha2_password` uses `RSA_PKCS1_PADDING`: https://github.com/mysql/mysql-server/blob/020025d9f277fd3283b8c75024f590857f58fbbd/sql-common/client_authentication.cc?utf8=%E2%9C%93#L497

Reply:
Or on the server-side, https://github.com/mysql/mysql-server/blob/020025d9f277fd3283b8c75024f590857f58fbbd/sql/auth/sql_authentication.cc?utf8=%E2%9C%93#L3245 vs https://github.com/mysql/mysql-server/blob/020025d9f277fd3283b8c75024f590857f58fbbd/sql/auth/sha2_password.cc?utf8=%E2%9C%93#L1081

Reply:
That was unexpected. Should I push the fix?

Reply:
Yes please; I think that's the only thing blocking this PR.